src/Features/Core/Portable/DisposeAnalysis/DisposeObjectsBeforeLosingScopeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,202 @@
+            description: new LocalizableResourceString(nameof(FeaturesResources.UseRecommendedDisposePatternDescription), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        public DisposeObjectsBeforeLosingScopeDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_disposeObjectsBeforeLosingScopeRule, s_useRecommendedDisposePatternRule), GeneratedCodeAnalysisFlags.None)
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                if (!DisposeAnalysisHelper.TryGetOrCreate(compilationContext.Compilation, out DisposeAnalysisHelper disposeAnalysisHelper))
+                {
+                    return;
+                }
+
+                // Avoid reporting duplicate diagnostics from interprocedural analysis.
+                var reportedLocations = new ConcurrentDictionary<Location, bool>();
+
+                compilationContext.RegisterOperationBlockAction(
+                    operationBlockContext => AnalyzeOperationBlock(operationBlockContext, disposeAnalysisHelper, reportedLocations));
+            });
+        }
+
+        private void AnalyzeOperationBlock(
+            OperationBlockAnalysisContext operationBlockContext,
+            DisposeAnalysisHelper disposeAnalysisHelper,
+            ConcurrentDictionary<Location, bool> reportedLocations)
+        {
+            // We are only intersted in analyzing method bodies that have at least one disposable object creation.
+            if (!(operationBlockContext.OwningSymbol is IMethodSymbol containingMethod) ||
+                !disposeAnalysisHelper.HasAnyDisposableCreationDescendant(operationBlockContext.OperationBlocks, containingMethod))
+            {
+                return;
+            }
+
+            // We can skip interprocedural analysis for certain invocations.
+            var interproceduralAnalysisPredicateOpt = new InterproceduralAnalysisPredicate(
+                skipAnalysisForInvokedMethodPredicateOpt: SkipInterproceduralAnalysis,
+                skipAnalysisForInvokedLambdaOrLocalFunctionPredicateOpt: null,
+                skipAnalysisForInvokedContextPredicateOpt: null);
+
+            // Compute dispose dataflow analysis result for the operation block.
+            if (disposeAnalysisHelper.TryGetOrComputeResult(operationBlockContext.OperationBlocks, containingMethod,
```

Comment:
since lifetime of IOperation is tied to SemanticModel, I guess this cache will get cleared when semantic model goes away? or will it strongly hold onto the IOperation?

Reply:
Dataflow analysis results are weakly keyed with IOperation using CWT [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowAnalysis.cs#L25)